06_gpu_and_ml/unsloth_finetune.py

```diff
@@ -0,0 +1,580 @@
+        eval_dataset = datasets.load_from_disk(dataset_cache_path / "eval")
+    else:
+        print(f"Downloading and processing dataset: {config.dataset_name}")
+
+        # Load and standardize the dataset format
+        dataset = datasets.load_dataset(config.dataset_name, split="train")
+        dataset = standardize_sharegpt(dataset)
+
+        # Split into training and evaluation sets with fixed seed for reproducibility
+        dataset = dataset.train_test_split(
+            test_size=1.0 - TRAIN_SPLIT_RATIO, seed=config.seed
+        )
+        train_dataset = dataset["train"]
+        eval_dataset = dataset["test"]
+
+        # Apply chat template formatting to convert conversations to text
+        print("Formatting datasets with chat template...")
+        train_dataset = train_dataset.map(
+            lambda examples: format_chat_template(examples, tokenizer),
+            batched=True,
+            num_proc=PREPROCESSING_WORKERS,
+            remove_columns=train_dataset.column_names,
+        )
+
+        eval_dataset = eval_dataset.map(
+            lambda examples: format_chat_template(examples, tokenizer),
+            batched=True,
+            num_proc=PREPROCESSING_WORKERS,
+            remove_columns=eval_dataset.column_names,
+        )
+
+        # Cache the processed datasets for future runs
+        print(f"Caching processed datasets to {dataset_cache_path}")
+        dataset_cache_path.mkdir(parents=True, exist_ok=True)
+        train_dataset.save_to_disk(dataset_cache_path / "train")
+        eval_dataset.save_to_disk(dataset_cache_path / "eval")
+
+        # Commit the dataset cache to the volume
+        dataset_cache_volume.commit()
+
+    return train_dataset, eval_dataset
+
+
+# ## Loading the pretrained model
+# We can't finetune without a pretarined model! Since these models are
+# fairly large, we don't want to download them from scratch for each training run.
+# We solve this by caching the weights in a Volume on download, and then loading
+# from the Volume on subsequent runs.
+def load_or_cache_model(config: "TrainingConfig", paths: dict):
+    model_cache_path = paths["model_cache"]
+
+    if model_cache_path.exists():
+        print(f"Loading cached model from {model_cache_path}")
+        model, tokenizer = FastLanguageModel.from_pretrained(
+            model_name=str(model_cache_path),
+            max_seq_length=config.max_seq_length,
+            dtype=None,
+            load_in_4bit=config.load_in_4bit,
+            load_in_8bit=config.load_in_8bit,
+        )
+    else:
+        print(f"Downloading and caching model: {config.model_name}")
+        model, tokenizer = FastLanguageModel.from_pretrained(
+            model_name=config.model_name,
+            max_seq_length=config.max_seq_length,
+            dtype=None,
+            load_in_4bit=config.load_in_4bit,
+            load_in_8bit=config.load_in_8bit,
+        )
```

Comment:
This pattern is no longer needed nor recommended because if we put `HF_HOME` in a Modal Volume, then the subsequent model instantiations will find the downloaded model and skip downloading it again. I suggest removing and just leaving:

```python
model_name=config.model_name,
```